## Homology determination between *Mus musculus* and *Danio rerio* analysis
### Introduction

This part of the work describes the analysis that was performed to continue
analyzing Tbx5 transcription factor and it's impact to heart cells'
regeneration.

The main purpose of the analysis was to determine the similarity of
*Mus musculus* and *Danio rerio* genomes by using various bioinformatical
tools in order to identify synteny blocks between beforementioned organisms.

The samples that were used in the analysis are described in the previous
analysis and can be viewed
[here](file:///home/daniele/Desktop/IV_course/I_semester/Kursinis_projektas/Tbx5_analysis_II/Programs/report/out_html/samples.html).
The results of the Tbx5 transcription factor abundance in *Mus musculus*
organism analysis can be found
[here](file:///home/daniele/Desktop/IV_course/I_semester/Kursinis_projektas/Tbx5_analysis_II/Programs/report/out_html/peaks_MM.html).


The analysis provides the following information:\
 **1.** 
 **2.** 

```{r Libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(pacman)
p_load(data.table, rtracklayer, ggplot2, ggthemes, plyranges, ggpubr, plotly,
    BRGenomics, reshape2, dplyr, gplots, Biostrings, scales, gtools, annotables,
    ChIPseeker, TxDb.Mmusculus.UCSC.mm10.knownGene, org.Mm.eg.db, org.Dr.eg.db,
    TxDb.Drerio.UCSC.danRer11.refGene, BSgenome.Drerio.UCSC.danRer11, car,
    BSgenome.Mmusculus.UCSC.mm10)
```

```{r Paths, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Declaring paths:
PROJECT <- paste0("/home/daniele/Desktop/IV_course/I_semester/",
                    "Kursinis_projektas/Tbx5_analysis_II/")
INPUTS <- paste0(PROJECT, "Inputs/")
INTER_FILES <- paste0(PROJECT, "Intermediate_data_files/")
FIGURES <- paste0(PROJECT, "Figures/")
RESULTS <- paste0(PROJECT, "Peak_annotations/")
GTF_FASTA <- paste0(PROJECT, "Seq_Annot/")
# nolint end
```

```{r Synteny_Blocks, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Reading a file that describes synteny between organisms:
synteny_blocks <- read.csv(paste0(INTER_FILES, "synteny_blocks.csv"), skip = 1)
samples <- read.csv(file = paste0(INTER_FILES, "sample_key.csv"))

# Mouse vs. Zebrafish
# Getting synteny block positions and chromosome number for mouse:
mouse_chr_pos <- synteny_blocks[, c(7:8, 6, 3:4, 2)]
colnames(mouse_chr_pos) <- c("startpos_MM", "endpos_MM", "chr_MM",
                             "startpos_DR", "endpos_DR", "chr_DR")
head(mouse_chr_pos)
# Creating a dataframe that stores data:
df_mm <- data.frame(chr = paste0("chr", c(mouse_chr_pos[, 3])),
       start = c(mouse_chr_pos[, 1]), end = c(mouse_chr_pos[, 2]), strand = "+")

# Checking whether the starting region position is bigger than end position:
for (i in 1:length(rownames(df_mm))) {
    if (df_mm[i, 2] > df_mm[i, 3]) {
        reverse1 = df_mm[i, 2]
        df_mm[i, 2] = df_mm[i, 3]
        df_mm[i, 3] = reverse1
        df_mm[i, 4] <- "-"
    }
}

# Creating a GRanges object for created dataframe:
syn_grange <- makeGRangesFromDataFrame(df_mm)

# Reading bigBed files:
bbfiles <- list.files(path = paste0(INPUTS, "BigBed/"), "*bb")
grl <- GRangesList()

# Declaring what chromosomes are analyzed:
chr_abr <- c(paste0("chr", 1:19), "chrX", "chrY")

# Creating GRanges objects and extracting certain chromosomes:
for (i in 1:length(bbfiles)) {
    grl[[i]] <- import(paste0(paste0(INPUTS, "BigBed/"), bbfiles[i])) %>%
                       dplyr::filter(seqnames %in% chr_abr)
    names(grl)[i] <- samples$Graph_names[samples$Filename == bbfiles[i]]
}
# nolint end
```

In comparison with the *Mus musculus* results, the number of fragments in
chromosomes apparently varies between the chromosomes. According to the diagram,
the greatest number of fragments has been detected in the 5, 7, 9 and 17
chromosomes (75, 74, 76, 74 and 76 fragments, respectively). The smallest number
of fragments has been detected in the 22nd chromosome (28 fragments).

### Synteny blocks percentage in sample peaks

```{r Synteny_percentage, fig.width = 6, fig.height = 5, fig.align = "center"}
# nolint start
indicators <- c("synteny_overlaps", "peak_count")
percentages <- c()
grl_reduced <- GRangesList()

# Creating an empty dataframe and naming it's columns:
df_final <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(df_final) <- c("names", "indic", "value", "percentage")

# Determining whether overlaps between synteny blocks and peaks exist:
for (i in 1:length(grl)) {
    total_peaks = length(grl[[i]])
    overlap_count = length(subsetByOverlaps(grl[[i]], syn_grange))
    values <- c(overlap_count, total_peaks)

    for (y in 1:length(indicators)) {
        row = c(samples$Graph_names[i], indicators[y], values[y],
                (overlap_count / total_peaks) * 100)
        df_final[nrow(df_final) + 1, ] <- row
    }

    grl_reduced[[i]] <- subsetByOverlaps(grl[[i]], syn_grange)
    names(grl_reduced)[i] <- names(grl[i])
}
# nolint end
```

```{r}
# nolint start
plot2 <- ggplot(df_final, aes(fill = indic, y = as.numeric(value), x = names)) + 
    geom_bar(width = 1, size = 0.2, colour = "#3f2704", stat = 'identity',
             position = position_dodge(0.8)) +
    geom_text(aes(label = ifelse(indic == "synteny_overlaps",
                                 paste0(round(as.numeric(percentage),
                                              digits = 2), "%"), ""),
              fontface = 2), vjust = -1.2, hjust = -0.5, size = 5) +
    labs(title = "", x = "", y = "") + 
    scale_fill_manual(values = c("#d37f02", "#f3bc47"),
                      labels = c("Pikų skaičius", "Synteny blokų skaičius")) +
    scale_y_continuous(labels = label_number(suffix = " K", scale = 1e-3)) +
    guides(fill = guide_legend(title = "Spalvų paaiškinimas")) +
    theme(panel.background = element_rect(fill = "#eeeef1",
                                          colour = "#4c0001"),
          panel.grid.major.y = element_line(colour = "#cab5b5", size = 0.3,
                                            linetype = "dashed"),
          panel.grid.minor.y = element_line(colour = "#cab5b5", size = 0.3,
                                            linetype = "dashed"),
          panel.grid.major.x = element_line(colour = "#cab5b5", size = 0.2,
                                            linetype = "longdash"),
          panel.grid.minor.x = element_line(colour = "#cab5b5", size = 0.2,
                                            linetype = "longdash"),
          axis.text.x = element_text(angle = 0, hjust = 1, vjust = 1, size = 11,
                                     face = "bold", color = "black"),
          axis.text.y = element_text(size = 11, face = "bold",
                                     color = "black"),
          axis.title.x = element_text(size = 2),
          axis.title.y = element_text(size = 16),
          plot.title = element_text(hjust = 0.5, face = "bold"),
          legend.title = element_text(face = "bold", size = 10),
          legend.text = element_text(size = 10)) +
    coord_flip()

# png(file = paste0(FIGURES, "Synteny_blocks_peaks.png"), width = 700)
# plot2
# dev.off()
plot2
# nolint end
```

### Peak annotation

```{r Peak_Annotations_MM, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
mm_known_genes <- TxDb.Mmusculus.UCSC.mm10.knownGene
sample_names <- c()
grl_synteny <- GRangesList()

# Annotating Mus musculus peaks that belong to synteny block:
for (object in 1:length(grl_reduced)) {
    name <- gsub(" ", "_", names(grl[object]))
    peak <- grl_reduced[[object]]
    seqlengths(peak) <- seqlengths(peak) - 10001
    peak_annotation <- annotatePeak(peak, tssRegion = c(-3000, 3000),
                         TxDb = mm_known_genes, annoDb = "org.Mm.eg.db")

    mm_annot <- as.data.frame(peak_annotation@anno)
    entrezids <- unique(mm_annot$geneId)
    entrez2gene <- grcm38 %>% filter(entrez %in% entrezids) %>%
                                dplyr::select(entrez, symbol)

    m <- match(mm_annot$geneId, entrez2gene$entrez)
    mm_annot <- cbind(mm_annot[, 1:14], gene_symbol = entrez2gene$symbol[m],
                        mm_annot[, 15:16])

    # Defining the same grl_smlr object that has two extra columns with
    # gene id and gene symbol:
    grl_synteny[[object]] <- mm_annot
    names(grl_synteny)[object] <- names(grl)[object]
}

write.table(grl_synteny, file = paste0(RESULTS, "/grl_synteny.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE)
# Next day: pateikti diagramą, vaizduojančią, kiek gali būti nustatyta genų,
#           priklausančių sintenijos blokui (unikalių genų kiekvienam mėginiui).
# nolint end
```

```{r Peak_Annotations_MM_Plot, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
genes_samples <- data.frame(matrix(ncol = 2, nrow = 0))
colnames(genes_samples) <- c("Sample", "Gene_count")

for (smpl in 1:length(grl_synteny)) {
    row = c(names(grl_synteny[smpl]),
            length(unique(grl_synteny[[smpl]]$gene_symbol)))
    genes_samples[nrow(genes_samples) + 1, ] <- row
}

# Diagram that shows how many unique genes were identified for each
# sample belonging to the synteny block:
plot3 <- ggplot(genes_samples, aes(x = Sample, y = as.numeric(Gene_count))) +
    geom_bar(stat = "identity", fill = "#cc8b12",
            width = .6, color = "#8d5c00") +
    labs(title = "", x = "", y = "Anotuotų pikų skaičius M. Musculus") +
    geom_text(aes(label = Gene_count), color = "#030101",
              size = 5,vjust = -1, fontface = 2) +
    theme(axis.text = element_text(size = 10, colour = "black"),
          axis.text.x = element_text(angle = 45, hjust = 1, size = 12,
                                     vjust = 1, face = "bold"),
          axis.text.y = element_text(size = 12, face = "bold"),
          axis.title.x = element_text(size = 14, colour = "black"),
          axis.title.y = element_text(size = 14, colour = "black"),
          panel.grid.major = element_line(color = "#eeeeee"),
          plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),
          panel.background = element_rect(fill = "#eeeef1", colour = "#3a1010"),
          panel.grid.major.y = element_line(colour = "#cab5b5",
                                            size = 0.3, linetype = "dashed"),
          panel.grid.minor.y = element_line(colour = "#cab5b5",
                                            size = 0.3, linetype = "dashed"),
          panel.grid.major.x = element_line(colour = "#cab5b5",
                                            size = 0.2, linetype = "longdash"),
          panel.grid.minor.x = element_line(colour = "#cab5b5",
                                            size = 0.2, linetype = "longdash"),
          legend.position = c(0.777, 0.788))

# png(file = paste0(FIGURES, "Synteny_blocks_peaks.png"), width = 700)
# plot3
# dev.off()
plot3
# nolint end
```

Čia bandom surasti tuos pačius genus Danio rerio genome.

```{r Danio_rerio_genes, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
# Getting positions of Danio rerio genes:
gene_ranges <- genes(TxDb.Drerio.UCSC.danRer11.refGene)

# Adding 'gene_symbol' column to gene_ranges GRange:
gene_ranges$gene_symbol <- mapIds(org.Dr.eg.db, keys = gene_ranges$gene_id,
                column = "SYMBOL", keytype = "ENTREZID", multiVals = "first")

# Getting synteny block positions and chromosome number for zebrafish:
fish_chr_pos <- synteny_blocks[, c(3:4, 2)]
colnames(fish_chr_pos) <- c("startpos", "endpos", "chr")

# Creating a dataframe that stores data:
df_dr <- data.frame(chr = paste0("chr", c(fish_chr_pos[, 3])),
       start = c(fish_chr_pos[, 1]), end = c(fish_chr_pos[, 2]), strand = "+")

for (i in 1:length(rownames(df_dr))) {
    if (df_dr[i, 2] > df_dr[i, 3]) {
        reverse1 = df_dr[i, 2]
        df_dr[i, 2] = df_dr[i, 3]
        df_dr[i, 3] = reverse1
        df_dr[i, 4] <- "-"
    }
}

# Creating a GRanges object for created dataframe:
syn_grange_dr <- makeGRangesFromDataFrame(df_dr)

dr_genes <- GRangesList()
commons <- GRangesList()

# Getting the genes that overlap with Mus musculus genes:
for (smpl2 in 1:length(grl_synteny)) {
    selected_genes <- tolower(gene_ranges$gene_symbol) %in%
                            tolower(unique(grl_synteny[[smpl2]]$gene_symbol))
    genes <- gene_ranges[selected_genes]

    dr_genes[[smpl2]] <- genes
    names(dr_genes)[smpl2] <- names(grl_synteny[smpl2])
}

# Creating an empty dataframe to store data about gene counts:
genes_synteny <- data.frame(matrix(ncol = 2, nrow = 0))
colnames(genes_synteny) <- c("Sample", "Gene_count")

# Finding genes that belong to synteny block:
for (smpl3 in 1:length(grl_synteny)) {
    common_genes <- subsetByOverlaps(dr_genes[[smpl3]], syn_grange_dr)
    row = c(names(grl_synteny[smpl3]), length(common_genes))
    genes_synteny[nrow(genes_synteny) + 1, ] <- row
    commons[[smpl3]] <- common_genes
    names(commons)[smpl3] <- names(grl)[smpl3]
}



head(mouse_chr_pos)

grl_mm <- GRangesList()
for (object in 1:length(grl)) {
    name <- gsub(" ", "_", names(grl[object]))
    peak <- grl[[object]]
    seqlengths(peak) <- seqlengths(peak) - 10001
    peak_annotation <- annotatePeak(peak, tssRegion = c(-3000, 3000),
                         TxDb = mm_known_genes, annoDb = "org.Mm.eg.db")

    mm_annot <- as.data.frame(peak_annotation@anno)
    entrezids <- unique(mm_annot$geneId)
    entrez2gene <- grcm38 %>% filter(entrez %in% entrezids) %>%
                                dplyr::select(entrez, symbol)

    m <- match(mm_annot$geneId, entrez2gene$entrez)
    mm_annot <- cbind(mm_annot[, 1:14], gene_symbol = entrez2gene$symbol[m],
                        mm_annot[, 15:16])

    # Defining the same grl_smlr object that has two extra columns with
    # gene id and gene symbol:
    grl_mm[[object]] <- mm_annot
    names(grl_mm)[object] <- names(grl)[object]
}


mm <- grl_mm[[1]]       # Mus musculus
dr <- gene_ranges       # Danio rerio

c <- as.data.frame(na.omit(dr))
d <- as.data.frame(na.omit(mm))
69069755  68683113
head(c)

intersecting_genes <- data.frame(matrix(ncol = 7, nrow = 0))
colnames(intersecting_genes) <- c("Gene", "Chr_MM", "Start_MM", "End_MM",
                                  "Chr_DR", "Start_DR", "End_DR")

for (i in 1:length(rownames(mouse_chr_pos))) {
    k <- mouse_chr_pos[664, ]
    a_dr <- na.omit(c[c$seqnames == paste0("chr", k$"chr_DR") &
              c$start >= k$"startpos_DR" & c$end <= k$"endpos_DR", ])
    a_mm <- na.omit(d[d$seqnames == paste0("chr", k$"chr_MM") &
              d$start >= k$"startpos_MM" & d$end <= k$"endpos_MM", ])
    if (length(rownames(a_dr)) != 0 & length(rownames(a_mm)) != 0 &
        length(intersect(tolower(a_dr[,"gene_symbol"]),
                         tolower(a_mm[,"gene_symbol"]))) != 0) {
        print(paste0("i value = ", i))
        # print(a_dr[,c("seqnames", "start", "end", "gene_symbol")])
        # print(a_mm[,c("seqnames", "start", "end", "gene_symbol")])

        l <- intersect(tolower(a_dr[,"gene_symbol"]),
                       tolower(a_mm[,"gene_symbol"]))
        dr_genes <- a_dr[a_dr$"gene_symbol" %in% l, c("seqnames", "start", "end", "gene_symbol")]

        dplc <- (a_mm[tolower(a_mm$"gene_symbol") %in% l,
                      c("seqnames", "start", "end", "gene_symbol")])
        # mm_genes <- dplc[!duplicated(dplc[ , "gene_symbol"]), c("seqnames", "start", "end", "gene_symbol")]
        # mm_genes$"gene_symbol" <- tolower(mm_genes$"gene_symbol")
        # print(dr_genes)
        # print(mm_genes)

        # merged <- merge(mm_genes, dr_genes, by = "gene_symbol")
        # intersecting_genes[nrow(intersecting_genes) + 1, ] <- merged

        write.table(dr_genes, file = paste0(INTER_FILES, "/dr_genes.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE,append=TRUE)
        write.table(dplc, file = paste0(INTER_FILES, "/dplc.txt"), sep = "\t",
            quote = FALSE, row.names = FALSE,append=TRUE)
    }
}


# nolint end
```

```{r Sequences, echo=FALSE, message=FALSE, warning=FALSE}
# nolint start
seq_dataset_dr <- c()

# Creating a sequence dataset that stores Mus musculus gene sequences:
mm_gr <- d
mm_gr <- mm_gr[mm_gr$start %in% intersecting_genes$"Start_MM" &
        mm_gr$end %in% intersecting_genes$"End_MM" &
        mm_gr$seqnames %in% paste0("chr", intersecting_genes$"Chr_MM"),
        c("seqnames", "start", "end", "gene_symbol")]

mm_gr <- makeGRangesFromDataFrame(mm_gr, keep.extra.columns=TRUE)


dr_gr <- c
dr_gr <- dr_gr[dr_gr$start %in% intersecting_genes$"Start_DR" &
        dr_gr$end %in% intersecting_genes$"End_DR" &
        dr_gr$seqnames %in% paste0("chr", intersecting_genes$"Chr_DR"),
        c("seqnames", "start", "end", "gene_symbol")]

dr_gr <- makeGRangesFromDataFrame(dplc, keep.extra.columns=TRUE)

gene_seq_mm <- getSeq(BSgenome.Mmusculus.UCSC.mm10, dr_gr)
names(gene_seq_mm) <- dr_gr$gene_symbol

gene_seq_dr <- getSeq(BSgenome.Drerio.UCSC.danRer11, dr_gr)
names(gene_seq_dr) <- dr_gr$gene_symbol


######################################################################

# Ar tai reiškia, kad sintenijos blokai leidžia surasti panašius
# regionus, tačiau nustatytose pozicijose gali jungtis tik labai
# mažas Tbx5 transkripcijos faktoriaus procentas?

# Patikrinti:
#   kaip yra su kiekvienu mėginiu?
#   sukurti grafikus, leidžiančius vizualizuoti procento skirtumus?

######################################################################



# as.character(gene_seq_mm[names(gene_seq_mm) == "Ccn6"])
# gene_seq_dr[names(gene_seq_dr) == "ccn6"]
# nchar(as.character(gene_seq_dr[1]))

mpwm <- read.table(paste0(INTER_FILES, "TBX5_MOUSE.H11MO.0.D.pwm"))
mpwm <- t(mpwm)
rownames(mpwm) <- c("A", "C", "G", "T")

gene_hits_dr <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(gene_hits_dr) <- c("Gene", "Hits", "Length")

# Determining the number of Tbx5 binding hits:
for (gene in 1:length(gene_seq_mm)) {
    hits <- countPWM(as.matrix(mpwm), as.character(gene_seq_mm[gene]),
                     min.score = "75%")
    # matchPWM(as.matrix(mpwm), as.character(seq_dataset[[gene]]]),
            #   min.score = "75%")
    row = c(names(gene_seq_mm[gene]), hits,
            nchar(as.character(gene_seq_mm[gene])))
    gene_hits_dr[nrow(gene_hits_dr) + 1, ] <- row
}

gene_hits_dr[order(gene_hits_dr$Gene),]


gene_hits_mm <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(gene_hits_mm) <- c("Gene", "Hits", "Length")

# Determining the number of Tbx5 binding hits:
for (gene in 1:length(gene_seq_mm)) {
    hits <- countPWM(as.matrix(mpwm), as.character(gene_seq_mm[gene]),
                     min.score = "75%")
    # matchPWM(as.matrix(mpwm), as.character(seq_dataset[[gene]]]),
            #   min.score = "75%")
    row = c(names(gene_seq_mm[gene]), hits,
            nchar(as.character(gene_seq_mm[gene])))
    gene_hits_mm[nrow(gene_hits_mm) + 1, ] <- row
}

gene_hits_mm[order(gene_hits_mm$Gene),]

# nolint end
```
